flink-connectors/flink-connector-gcp-pubsub/src/main/java/org/apache/flink/streaming/connectors/gcp/pubsub/table/PubsubDynamicSource.java

```diff
@@ -0,0 +1,80 @@
+package org.apache.flink.streaming.connectors.gcp.pubsub.table;
+
+import org.apache.flink.annotation.Internal;
+import org.apache.flink.api.common.serialization.DeserializationSchema;
+import org.apache.flink.streaming.connectors.gcp.pubsub.PubSubSource;
+import org.apache.flink.table.connector.ChangelogMode;
+import org.apache.flink.table.connector.format.DecodingFormat;
+import org.apache.flink.table.connector.source.DynamicTableSource;
+import org.apache.flink.table.connector.source.ScanTableSource;
+import org.apache.flink.table.connector.source.SourceFunctionProvider;
+import org.apache.flink.table.data.RowData;
+import org.apache.flink.table.types.DataType;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+
+/** PubSub-backed {@link ScanTableSource}. */
+@Internal
+public class PubsubDynamicSource implements ScanTableSource {
+
+    private static Logger logger = LoggerFactory.getLogger(PubsubDynamicSource.class);
+
+    /** Name of the PubSub project backing this table. */
+    private final String project;
+    /** Name of the PubSub subscription backing this table. */
+    private final String subscription;
+    /** Scan format for decoding records from PubSub. */
+    private final DecodingFormat<DeserializationSchema<RowData>> decodingFormat;
+    /** Data type that describes the final output of the source. */
+    private final DataType producedDataType;
+
+    public PubsubDynamicSource(
+            String project,
+            String subscription,
+            DecodingFormat<DeserializationSchema<RowData>> decodingFormat,
+            DataType producedDataType) {
+
+        this.project = project;
+        this.subscription = subscription;
+        this.decodingFormat = decodingFormat;
+        this.producedDataType = producedDataType;
+    }
+
+    @Override
+    public ChangelogMode getChangelogMode() {
+        return ChangelogMode.insertOnly();
+    }
+
+    @Override
+    public ScanRuntimeProvider getScanRuntimeProvider(ScanContext runtimeProviderContext) {
+        // create runtime classes that are shipped to the cluster
+
+        final DeserializationSchema<RowData> deserializer =
+                decodingFormat.createRuntimeDecoder(runtimeProviderContext, producedDataType);
+
+        try {
+            PubSubSource<RowData> source =
+                    PubSubSource.newBuilder()
+                            .withDeserializationSchema(deserializer)
+                            .withProjectName(project)
+                            .withSubscriptionName(subscription)
```

Comment:
(We'll need to think about how to provide credentials (not through environment variables) as well, I think, but as a first version we can leave this as is.)

Reply:
I think, most of the time,  when you read from PubSub your cluster exists in GCP and you probably configure the state with GCS, so you already define the credentials.
Same for Kinesis, your cluster probably deployed in AWS with s3 state and you already define the credentials.(I can't see in Kinesis that they get credentials... maybe it should be addressed  also there)

This can be relevant when you run in one cloud, let's say AWS ,and want to read data from GCP PubSub and vice versa.. or you run  on premises cluster that reads data from kinesis/PubSub.. for those cases yes. we need to find a way to read credentials in some secured way, because tables can be done by external teams (self service)